File: ruby/875.koko-eating-bananas.rb
# @param {Integer[]} piles
# @param {Integer} h
# @return {Integer}
def min_eating_speed(piles, h)
  left = 1
  right = piles.max
  current_k = right

  while left < right
    speed = ((left + right) / 2).floor

    sum_time = piles.inject(0) do |time, pile|
      time += (pile / speed.to_f).ceil
    end

    if sum_time <= h
      right = speed
      current_k = current_k <= speed ? current_k : speed
    elsif left = speed + 1
    end
  end

  current_k
end

File: rust/src/875.koko-eating-bananas.rs
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        use std::cmp::Ordering;
        let mut left = 1;
        let mut right = *piles.iter().max().unwrap();
        let mut current_k = *piles.iter().max().unwrap();

        while left < right {
            let speed = (left + right) / 2;

            let sum_time = piles.iter().map(|&x| (x + speed - 1) / speed).sum::<i32>();

            match sum_time.cmp(&h) {
                Ordering::Equal => {
                    right = speed;
                    current_k = std::cmp::min(speed, current_k);
                }
                Ordering::Less => {
                    right = speed;
                    current_k = std::cmp::min(speed, current_k);
                }
                Ordering::Greater => left = &speed + 1,
            }
        }
        current_k
    }
}
